src/source-dpdk.c

```diff
@@ -0,0 +1,494 @@
+#endif
+}
+
+static void DPDKReleasePacket(Packet *p)
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet) */
+    if ((p->dpdk_v.copy_mode != DPDK_COPY_MODE_NONE) && !PKT_IS_PSEUDOPKT(p) &&
+            p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS && !PACKET_TEST_ACTION(p, ACTION_DROP)) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            rte_pktmbuf_free(p->dpdk_v.mbuf);
+            p->dpdk_v.mbuf = NULL;
+        }
+    } else {
+        rte_pktmbuf_free(p->dpdk_v.mbuf);
+        p->dpdk_v.mbuf = NULL;
+    }
+
+    PacketFreeOrRelease(p);
+}
+
+/**
+ *  \brief Main DPDK reading Loop function
+ */
+TmEcode ReceiveDPDKLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+    Packet *p;
+    uint16_t nb_rx;
+    time_t last_dump = 0;
+    time_t current_time;
+
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    PacketPoolWait();
+    while (1) {
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogInfo("Stopping Suricata!");
+            DPDKDumpCounters(ptv);
+            break;
+        }
+
+        nb_rx = rte_eth_rx_burst(ptv->port_id, ptv->queue_id, ptv->received_mbufs, BURST_SIZE);
+        if (unlikely(nb_rx == 0)) {
+            continue;
+        }
+
+        ptv->pkts += (uint64_t)nb_rx;
+        for (int i = 0; i < nb_rx; i++) {
+            p = PacketGetFromQueueOrAlloc();
+            if (unlikely(p == NULL)) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to get Packet Buffer for DPDK mbuf!");
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i, i);
+                SCReturnInt(EXIT_FAILURE);
+            }
+            PKT_SET_SRC(p, PKT_SRC_WIRE);
+            p->datalink = LINKTYPE_ETHERNET;
+            if (ptv->checksum_mode == CHECKSUM_VALIDATION_DISABLE) {
+                p->flags |= PKT_IGNORE_CHECKSUM;
+            }
+
+            DPDKSetTimeval(&p->ts);
+            p->dpdk_v.mbuf = ptv->received_mbufs[i];
+            p->ReleasePacket = DPDKReleasePacket;
+            p->dpdk_v.copy_mode = ptv->copy_mode;
+            p->dpdk_v.out_port_id = ptv->out_port_id;
+            p->dpdk_v.out_queue_id = ptv->queue_id;
+
+            if (PacketSetData(p, rte_pktmbuf_mtod(p->dpdk_v.mbuf, uint8_t *),
+                        rte_pktmbuf_pkt_len(p->dpdk_v.mbuf)) == -1) {
+                TmqhOutputPacketpool(ptv->tv, p);
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i - 1, i + 1);
+                SCReturnInt(EXIT_FAILURE);
+            }
+
+            if (TmThreadsSlotProcessPkt(ptv->tv, ptv->slot, p) != TM_ECODE_OK) {
+                TmqhOutputPacketpool(ptv->tv, p);
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i - 1, i + 1);
```

Comment:
not clear why the mbuf array of n packets are getting dumped, at line 277 you are accessing per mbuf, so should you not free the current mbuf. rework or else comment and explain the logic

Reply:
@lukashino 

I added comments to the code where DPDKFreeMbufArray is used to explain the logic
[VV] this is next PR you have added and not in current code right?

it serves to free all unprocessed mbufs by Suricata (those mbufs which were not assigned to Suricata packets (line 278)).
[VV] I am not clear with this statement, according to your logic each worker get packet from port-queue pair. This is by `rx_burst`. So you can only free what you get as rx_burst.


When an error occurs, mbufs already assigned to these packets are freed by Suricata exit procedures. However, mbufs that were not transferred to Suricata would be lost.
[VV] Most of DPDK supported NIC implement Tail Drops, so any packets not obtained in rx_burst is present in NIC HW queue, any new packets coming to NIC will be dropped by NIC. what do you mean by ` not transferred to Suricata would be lost.` because in your loop you called rx_burst. these now in Suricata context.

 Mbuf array (where all fetched mbufs are stored) is declared in the ReceiveDPDKLoop. That is the reason, why the remaining packets are freed before the return statement.
[VV] I think you are talking about `ctrol + c` or exit or segfautl case. If that is right, you can simply ignore the packets if you are primary mode. 

If you are secondary mode, it is duty of primary to identify the loss of secondary and reset the port if there any.

Reply:
> I am not clear with this statement, according to your logic each worker get packet from port-queue pair. This is by rx_burst. So you can only free what you get as rx_burst. 

Well yes, there was probably a slight misunderstanding, DPDKFreeMbufArray frees the array of unprocessed mbufs by Suricata in the given worker (what I received in the current iteration of ReceiveLoop)

> what do you mean by not transferred to Suricata would be lost. because in your loop you called rx_burst. these now in Suricata context.

That's the thing. The mbufs are in Suricata context - they are not freed by the NIC. BUT, the mbufs are not assigned to Suricata packet structures. Suricata during exit procedures calls `p->ReleasePacket` function on all packets in its packet pool. To assign mbuf to the Suricata packet structure, you first need to obtain it (`PacketGetFromQueueOrAlloc`) and then properly initialize it. 

If initialization fails on one packet from the array then Suricata calls p->ReleasePacket function on packets before that packet. That call frees the mbufs. I need to clear the mbufs after the failed mbuf because Suricata does not really know about them yet.